strings-csharp/SubstringInString/SubstringInString/SubstringSearchMethods.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System.Text.RegularExpressions;
+
+public static partial class SubstringSearchMethods
+{
+    public static List<int> FindAllIndexesWithSubstring(string input, string search)
+    {
+        List<int> indexes = new();
+
+        for (var i = 0; i <= input.Length - search.Length; i++)
+        {
+            if (input.Substring(i, search.Length) == search)
+            {
+                indexes.Add(i);
+            }
+        }
+
+        return indexes;
+    }
+
+    public static void PrintResult(List<int> result)
+    {
+        Console.WriteLine("Indexes where the substring is found: " + string.Join(", ", result));
+    }
+
+    public static List<int> FindAllIndexesWithSpan(string input, string search)
+    {
+        List<int> indexes = new();
+
+        var inputSpan = input.AsSpan();
+        var searchSpan = search.AsSpan();
+
+        for (int i = 0; i <= input.Length - search.Length; i++)
+        {
+            if (inputSpan.Slice(i, search.Length).SequenceEqual(searchSpan))
+            {
+                indexes.Add(i);
+            }
+        }
+
+        return indexes;
+    }
+
+    public static List<int> FindAllIndexesWithIndexOf(string input, string search)
+    {
+        List<int> indexes = new();
+        var startIndex = 0;
+
+        do
+        {
+            startIndex = input.IndexOf(search, startIndex);
+            if (startIndex != -1)
+            {
+                indexes.Add(startIndex);
+                startIndex++;
+            }
+        }
+        while (startIndex != -1);
+
+        return indexes;
+    }
+
+    [GeneratedRegex("ip", RegexOptions.IgnoreCase, "en-US")]
+    private static partial Regex GeneratedRegex();
+    public static List<int> FindAllIndexesWithRegex(string input, string search)
+    {
+        List<int> indexes = new();
+        var regex = GeneratedRegex();
+
+        foreach (Match match in regex.Matches(input))
+        {
+            indexes.Add(match.Index);
+        }
+
+        return indexes;
+    }
```

Comment:
Since we are taking the search value as an argument, we can't use a source generated regex. This method signature could be changed to take a Regex though so that in the case of the benchmark especially, we can pass in the same regex for each execution and won't be generating it every time. We could then still use a source generated regex, just the source generation would now take place outside of this class.